File: src/main/java/io/github/binaryfoo/decoders/apdu/InternalAuthenticateAPDUDecoder.java
package io.github.binaryfoo.decoders.apdu;

import io.github.binaryfoo.DecodedData;
import io.github.binaryfoo.decoders.DecodeSession;

public class InternalAuthenticateAPDUDecoder implements CommandAPDUDecoder {
    @Override
    public APDUCommand getCommand() {
        return APDUCommand.InternalAuthenticate;
    }

    @Override
    public DecodedData decode(String input, int startIndexInBytes, DecodeSession session) {
        int length = Integer.parseInt(input.substring(8, 10), 16);
        String data = input.substring(10, 10 + length * 2);
        return DecodedData.primitive("C-APDU: Internal Authenticate", data, startIndexInBytes, startIndexInBytes + 6 + length);
    }
}

File: src/main/java/io/github/binaryfoo/decoders/apdu/InternalAuthenticateAPDUDecoder.kt
package io.github.binaryfoo.decoders.apdu

import io.github.binaryfoo.DecodedData
import io.github.binaryfoo.EmvTags
import io.github.binaryfoo.decoders.DecodeSession
import io.github.binaryfoo.decoders.PopulatedDOLDecoder
import java.util.logging.Logger

class InternalAuthenticateAPDUDecoder : CommandAPDUDecoder {
  override fun getCommand(): APDUCommand {
    return APDUCommand.InternalAuthenticate
  }

  override fun decode(input: String, startIndexInBytes: Int, session: DecodeSession): DecodedData {
    val length = Integer.parseInt(input.substring(8, 10), 16)
    val data = input.substring(10, 10 + length * 2)
    val ddolValues = decodeDDOLElements(session, data, startIndexInBytes + 5)
    return DecodedData.constructed("C-APDU: Internal Authenticate", data, startIndexInBytes, startIndexInBytes + 6 + length, ddolValues)
  }

  private fun decodeDDOLElements(session: DecodeSession, populatedDdol: String, startIndexInBytes: Int): List<DecodedData> {
    val cdol = session[EmvTags.DDOL]
    if (cdol != null) {
      try {
        return PopulatedDOLDecoder().decode(cdol, populatedDdol, startIndexInBytes, session)
      } catch (e: Exception) {
        LOG.throwing(GenerateACAPDUDecoder::class.java.simpleName, "decodeDDOLElements", e)
      }

    }
    return listOf()
  }

  companion object {
    private val LOG = Logger.getLogger(InternalAuthenticateAPDUDecoder::class.java.name)
  }

}
